ark/jsonschema/object.ts

```diff
@@ -0,0 +1,336 @@
+					})
+				}
+			}
+		)
+	})
+
+	// NB: We don't validate compatability of schemas for overlapping patternProperties
+	// since getting the intersection of regexes is inherenetly difficult.
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.entries(data).forEach(([dataKey, dataValue]) => {
+			patternProperties.forEach(([pattern, parsedJsonSchema]) => {
+				if (pattern.test(dataKey) && !parsedJsonSchema.allows(dataValue)) {
+					errors.push(
+						ctx.reject({
+							actual: dataValue,
+							expected: `${parsedJsonSchema.description} as property ${dataKey} matches patternProperty ${pattern}`
+						})
+					)
+				}
+			})
+		})
+		return errors.length === 0
+	}
+}
+
+const parsePropertyNames = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	if (!("propertyNames" in jsonSchema)) return
+
+	const propertyNamesValidator = innerParseJsonSchema.assert(
+		jsonSchema.propertyNames
+	)
+
+	if (
+		"domain" in propertyNamesValidator.json &&
+		propertyNamesValidator.json.domain !== "string"
+	) {
+		ctx.reject({
+			path: ["propertyNames"],
+			actual: `a schema for validating a ${propertyNamesValidator.json.domain as string}`,
+			expected: "a schema for validating a string"
+		})
+	}
+
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.keys(data).forEach(key => {
+			if (!propertyNamesValidator.allows(key)) {
+				errors.push(
+					ctx.reject({
+						message: `property ${key} doesn't adhere to the propertyNames schema of ${propertyNamesValidator.description}`
+					})
+				)
+			}
+		})
+		return errors.length === 0
+	}
+}
+
+const parseRequiredAndOptionalKeys = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	const optionalKeys: string[] = []
+	const requiredKeys: string[] = []
+	if ("properties" in jsonSchema) {
+		if ("required" in jsonSchema) {
+			if (jsonSchema.required.length !== new Set(jsonSchema.required).size) {
+				ctx.reject({
+					expected: "an array of unique strings",
+					path: ["required"]
+				})
+			}
+
+			for (const key of jsonSchema.required) {
+				if (key in jsonSchema.properties) requiredKeys.push(key)
+				else {
+					ctx.reject({
+						actual: key,
+						expected: "a key in the 'properties' object",
+						path: ["required"]
+					})
+				}
+			}
+			for (const key in jsonSchema.properties)
+				if (!jsonSchema.required.includes(key)) optionalKeys.push(key)
+		} else {
+			// If 'required' is not present, all keys are optional
+			optionalKeys.push(...Object.keys(jsonSchema.properties))
+		}
+	} else if ("required" in jsonSchema) {
+		ctx.reject({
+			actual:
+				"an object JSON Schema with 'required' array but no 'properties' object",
+			expected: "a valid object JSON Schema"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: innerParseJsonSchema.assert(jsonSchema.properties![key]).internal
```

Comment:
In general, instead of needing to reassert this later, once you validate a property is present, I'd assign it to its own variable like `properties` right away.


Although in this case it's not clear from the logic in the function why `properties` would always be defined. If that's something external where we know either `required` or `properties` will be defined when we call this function or something, it's best to refactor it in a way where those external assumptions don't exist.

Reply:
> In general, instead of needing to reassert this later, once you validate a property is present, I'd assign it to its own variable like `properties` right away.

It's not clear to me what you mean by this. Wdym by "needing to reassert this later", and wdym by "assign it to its own variable"? I think it's not clear to me what "it" is.

> Although in this case it's not clear from the logic in the function why `properties` would always be defined.

In order for L181 (correct line number at the time of your comment) to be run, `optionalKeys` cannot be empty (since the line is within `optionalKeys.map`), and we only ever append to `optionalKeys` under the `"properties" in jsonSchema` path -- thus if `optionalKeys` isn't empty, then `jsonSchema.properties` must be defined.